src/berry.c

```diff
@@ -1849,3 +2403,210 @@ void SetBerryTreesSeen(void)
+};
+
+static u8 GetMutationOutcome(u8 berry1, u8 berry2)
+{
+    u8 i;
+    for(i = 0; i < ARRAY_COUNT(sBerryMutations); i++)
+    {
+        if ((sBerryMutations[i][0] == berry1 && sBerryMutations[i][1] == berry2)
+          ||(sBerryMutations[i][0] == berry2 && sBerryMutations[i][1] == berry1))
+            return (i + 1);
+    }
+    return 0;
+}
+
+static u8 TryForMutation(u8 berryTreeId, u8 berry)
+{
+    u8 i, j, mulch;
+    s16 x1, x2, y1, y2;
+
+    // Get location of current tree
+    for (i = 0; i < OBJECT_EVENTS_COUNT; i++)
+    {
+        if (gObjectEvents[i].trainerRange_berryTreeId == berryTreeId && gObjectEvents[i].movementType == MOVEMENT_TYPE_BERRY_TREE_GROWTH)
+            break;
+    }
+    if (i == OBJECT_EVENTS_COUNT)
+        return 0;
+
+    x1 = gObjectEvents[i].currentCoords.x;
+    y1 = gObjectEvents[i].currentCoords.y;
+
+    mulch = GetMulchByBerryTreeId(GetObjectEventBerryTreeId(i));
+
+    // Try mutation for each adjacent tree
+    for (j = 0; j < OBJECT_EVENTS_COUNT; j++)
+    {
+        if (gObjectEvents[j].active && gObjectEvents[j].movementType == MOVEMENT_TYPE_BERRY_TREE_GROWTH && GetStageByBerryTreeId(GetObjectEventBerryTreeId(j)) != BERRY_STAGE_NO_BERRY && j != i)
+        {
+            x2 = gObjectEvents[j].currentCoords.x;
+            y2 = gObjectEvents[j].currentCoords.y;
+            if (Random() % 100 < (BERRY_MUTATION_CHANCE * (mulch == ITEM_TO_MULCH(ITEM_SURPRISE_MULCH) || mulch == ITEM_TO_MULCH(ITEM_AMAZE_MULCH))) && (
+                (x1 == x2 && y1 == y2 - 1) ||
+                (x1 == x2 && y1 == y2 + 1) ||
+                (x1 == x2 - 1 && y1 == y2) ||
+                (x1 == x2 + 1 && y1 == y2)))
+                return GetMutationOutcome(berry, gSaveBlock1Ptr->berryTrees[GetObjectEventBerryTreeId(j)].berry);
+        }
+    }
+    return 0;
+}
+#endif
+
+struct TreeMutationBitfield {
+  u8 a: 2;
+  u8 b: 2;
+  u8 unused: 4;
+};
+
+union TreeMutation {
+  u8 value;
+  struct TreeMutationBitfield asField;
+};
+
+static u8 GetTreeMutationValue(u8 id)
+{
+#if OW_BERRY_MUTATIONS
+    struct BerryTree *tree = GetBerryTreeInfo(id);
+    union TreeMutation myMutation;
+    if (tree->stopGrowth) // Pre-generated trees shouldn't have mutations
+        return 0;
+    myMutation.asField.a = tree->mutationA;
+    myMutation.asField.b = tree->mutationB;
+    return sBerryMutations[myMutation.value - 1][2];
+#else
+    return 0;
+#endif
+}
+
+static void SetTreeMutations(u8 id, u8 berry)
+{
+#if OW_BERRY_MUTATIONS == TRUE
+    struct BerryTree *tree = GetBerryTreeInfo(id);
+    union TreeMutation myMutation;
+
+    myMutation.value = TryForMutation(id, berry);
+    tree->mutationA = myMutation.asField.a;
+    tree->mutationB = myMutation.asField.b;
+#endif
+}
+
+static u16 GetBerryPestSpecies(u8 berryId)
+{
+#if OW_BERRY_PESTS == TRUE
+    const struct Berry *berry = GetBerryInfo(berryId);
+    switch(berry->color)
+    {
+        case BERRY_COLOR_RED:
+            return SPECIES_LEDYBA;
```

Comment:
Shouldn't this block check if the families are enabled?

Reply:
Correct, this PR was made before species refactor was pushed. I'll add the checks shortly